plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/GitHook.kt

```diff
@@ -0,0 +1,152 @@
+package org.jlleitschuh.gradle.ktlint
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Project
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+import org.intellij.lang.annotations.Language
+import java.io.File
+import javax.inject.Inject
+
+const val FILTER_INCLUDE_PROPERTY_NAME = "internalKtlintGitFilter"
+
+@Language("Bash")
+internal val shShebang = """
+#!/bin/sh
+set -e
+
+""".trimIndent()
+
+internal const val startHookSection = "######## KTLINT-GRADLE HOOK START ########\n"
+internal const val endHookSection = "######## KTLINT-GRADLE HOOK END ########\n"
+
+@Language("Bash")
+internal fun generateGitHook(taskName: String) = """
+
+CHANGED_FILES="${'$'}(git --no-pager diff --name-status --no-color --cached | awk '$1 != "D" && $2 ~ /\.kts|\.kt/ { print $2}')"
+
+if [[ -z "${'$'}CHANGED_FILES" ]]; then
+    echo "No Kotlin staged files."
+    exit 0
+fi;
+
+echo "Running ktlint over these files:"
+echo "${'$'}CHANGED_FILES"
+
+./gradlew --quiet $taskName -P$FILTER_INCLUDE_PROPERTY_NAME="${'$'}CHANGED_FILES"
```

Comment:
Are there any weird file names that would cause this to blow up? Eg. File names with spaces in them?

Reply:
Having spaces in file names is ok, as this property is split using `'\n'` (end of line) character.

And if there are such file names - I would not do here premature optimization, but would wait for issues. Githook install tasks have hook-update mechanism on new run, so it will be easy to update git hook on new version.